DOM-based measurements

as-cii opened the pull request:
Supersedes and closes #6083.
Supersedes and closes #7945.
## A Brief Flashback

Over the past month we have evaluated several strategies to attack the problem, among which:
#### iframe

Build the lines inside an `<iframe>` element and measure from them.

**Pros**: measurements happen in a separate context, completely decoupled from the current DOM. Some optimization can, therefore, be applied: skinnier DOM, less stylesheet, and so on.
**Cons**: building the lines twice has a cost which is higher than the optimizations we tried out. Moreover, we cannot alter the original DOM structure as, otherwise, measurements could be wrong. 
#### Canvas

Have a canvas element and use the `measureText` function to provide all the metrics we need to position visual elements.

**Pros**: measurements happen in a separate context and the overhead is minimal.
**Cons**: CSS typography is not fully supported (e.g. ligatures/contextual alternates get screwed up) and the risk of reporting corrupted measurements is very high.
#### No Measurements

Instead of positioning decorations, cursors, etc. with absolute coordinates, embed them inside each line and let the browser figure out all the metrics.

**Pros**: no need for measurements and more semantical HTML. In addition, this enriches the power of styling customizations (e.g. the user has the possibility to style a piece of text surrounded by a decoration).
**Cons**: worse performance, because the DOM gets more complicated and nesting increases. Overlays need to be displayed on top of lines node, which forces us to employ tricks to let them appear in front of everything. 
## The Present

In this PR we have implemented DOM-based measurements. The idea here is to append the real lines to the DOM, measure from them, and render everything else that needs measurements. 

Every time `TextEditorElement::pixelPositionForScreenPosition` or `TextEditorElement::screenPositionForPixelPosition` is called, we trigger a reflow (a bit like the DOM does when you try to read from it when its information is stale).

In `TextEditorPresenter` we update the DOM with all lines we'll need to measure before computing any state, in order to avoid multiple reflows caused by reading from line nodes more than once.

What's missing:
- [x] Autoscroll when reaching the right end of the editor.
- [x] Implement `screenPositionForPixelPosition` in `LinesYardstick`.
- [x] Fix remaining specs.
- [x] Avoid using `NodeIterator` as it is slow. Cache text nodes and measure those instead.
- [x] Evaluate if caching already computed positions yields some improvements.
- [x] Remove measurement in `TextEditorComponent::readAfterUpdateSync`.
- [x] Ensure we have no significant regression performance-wise.
- [x] Investigate why Hasklig behaves oddly.
- [x] Fix bug where the cursor disappears when moving it left/right.
- [x] :arrow_up: autocomplete-plus.
- [x] :arrow_up: wrap-guide.

Feedback is super welcome and appreciated! Thanks! :bow: 

/cc: @nathansobo @atom/feedback for :eyes:

as-cii committed:
:art: Refactor ::updateTilesState to a composed method

as-cii committed:
Compute the state for the longest line on screen

as-cii committed:
Render the longest screen row without painting it

as-cii committed:
Differentiate between gutter and lines tiles

...and fix specs as well. :green_heart:

as-cii committed:
Decouple horizontal dimensions computation from vertical ones

The former needs measurement and, therefore, depends on the latter.

as-cii committed:
Defer state building in TextEditorPresenter

as-cii committed:
:fire: Avoid mounting the container in ctor

...as it will be mounted anyways when updateSync gets called

as-cii committed:
Add LinesYardstick

...and create a MockLineNodesProvider for testing purposes

as-cii committed:
:green_heart: Fix TextEditorPresenter specs

...so that they play nicely with LinesYardstick

as-cii committed:
Trigger ::onWillMeasure before measuring anything

as-cii committed:
Integrate LinesYardstick with TextEditorComponent

as-cii committed:
Merge branch 'master' into as-double-reflow-measurements

as-cii committed:
:fire:

benogle commented on spec/text-editor-presenter-spec.coffee:
```diff
@@ -832,11 +889,46 @@ describe "TextEditorPresenter", ->
+
+          expectValues presenter.getState().content.tiles[6], {
+            visibility: "hidden"
+          }
```
can we get rid of these curlies?

nathansobo replied:
Does it really matter? Looking forward to ES6. :wink:

as-cii replied:
I am okay with removing them but this way the code feels a bit more consistent (i.e. the other specs use `expectValues` with curly brackets). 

What do you think if we just keep them for now?

maxbrunsfeld replied:
:+1:

benogle commented:
Excited for this!

as-cii committed:
:art:

as-cii committed:
:green_heart: Use tokenizedLine.id to refer to lines

as-cii committed:
Merge branch 'master' into as-double-reflow-measurements

as-cii committed:
Avoid calling getBoundingClientRect twice for each line

as-cii committed:
Merge branch 'master' into as-double-reflow-measurements

as-cii committed:
:racehorse: Cache positions

nathansobo commented on spec/text-editor-presenter-spec.coffee:
```diff
@@ -209,10 +220,36 @@ describe "TextEditorPresenter", ->
         presenter = buildPresenter()
         expectNoStateUpdate presenter, -> presenter.getState()
 
+      it "triggers ::onWillMeasure events before computing any state that needs measurement", ->
```
Think I asked this question on another PR with a similar method but... shouldn't this method be called like `onWillNeedMeasurements` or something?

as-cii replied:
Sounds good! :+1:

as-cii committed:
:green_heart:

as-cii committed:
:art: Rename to ::onWillNeedMeasurements

as-cii committed:
Merge branch 'master' into as-double-reflow-measurements

# Conflicts:
#	src/text-editor-presenter.coffee

as-cii committed:
Improve LinesYardstick design

We have shifted the responsibility of orchestrating state updates and
measurements to the yardstick. The presenter still needs to be updated to make
use of these new capabilities.

as-cii committed:
Revert changes to the presenter and the component

as-cii committed:
Start refactoring the presenter to accommodate LinesYardstick

as-cii committed:
:fire: Avoid computing top for line and line numbers

as-cii committed:
:bug: Avoid creating a useless extra tile

as-cii committed:
Revert previous commit

as-cii committed:
Use arrays instead of while loops

as-cii committed:
Implement ::getStateForMeasurements(screenRows)

as-cii committed:
Start integrating the yardstick

as-cii committed:
Redesign LinesYardstick

as-cii committed:
wip

as-cii committed:
Rename to TextEditorPresenter::invalidateState

as-cii committed:
:green_heart: Fix presenter specs

as-cii committed:
Merge branch 'master' into as-double-reflow-measurements

as-cii committed:
Reset only lines state

as-cii committed:
Revert "Reset only lines state"

This reverts commit 39e8920c72232f74a37c60e26b08b37c12964def.

as-cii committed:
:art: Some refactoring before updating from master

as-cii committed:
Merge branch 'master' into as-double-reflow-measurements

# Conflicts:
#	src/text-editor-presenter.coffee

as-cii committed:
Fix linting errors

as-cii committed:
Fix soft-wrapping spec

as-cii committed:
Merge branch 'master' into as-double-reflow-measurements

as-cii committed:
Remove yardstick to start with a clean slate

as-cii committed:
Decouple pending logical scroll positions

as-cii committed:
Split presenter state in pre/post measurement

as-cii committed:
Always render the longest screen row

as-cii committed:
Pass LinesYardstick to TextEditorPresenter

as-cii committed:
Temporarily disable caching from yardstick

This will enable us to avoid worrying about cache invalidation while integrating
`LinesYardstick` into `TextEditorPresenter`.

as-cii committed:
Use LinesYardstick in pixelPositionForScreenPosition

as-cii committed:
:green_heart: Fix TextEditorPresenter specs

as-cii committed:
Implement ::screenPositionForPixelPosition in LinesYardstick

as-cii committed:
:fire: :tada: Remove scoped character width

as-cii committed:
:green_heart:

as-cii committed:
Merge branch 'master' into as-double-reflow-measurements

# Conflicts:
#	src/text-editor-presenter.coffee

as-cii committed:
:racehorse: Compute lines state only once

as-cii committed:
:bug: Move accidentally leftover statement

as-cii committed:
:racehorse: Remove AcceptFilter

...so that Chrome doesn't need to switch to user space when iterating over line
nodes.

as-cii committed:
:racehorse: Cache pixel positions

as-cii committed:
Avoid to call ::prepareScreenRows explicitly

as-cii commented:
Travis is playing a bit strangely with the network today, but specs should be :green_heart: now and I feel like this is ready to be reviewed. :tada: 

I'll have a second look with fresh :eyes: tomorrow, but I feel pretty good about the code changes (I am sure there's room for improvement, though, especially in `TextEditorPresenter`). I have this feeling sometimes that the presenter wants to be split into smaller objects, but I haven't come up with a cohesive design yet (of course this is outside the scope of this PR).

This definitely needs to be tested out and I'd like to make sure we have no evident regression (e.g. :bug:s  or slowdowns). I am so happy that after this PR we should be able to :ship: subpixel font scaling pretty smoothly! :sparkles: 

Some more :eyes: are most welcome on this, and I'd like to invite @atom/feedback to try this out as well: ideally you shouldn't notice any change from **master**, exception made for ligatures and kerning which should now work properly.

-------------------------------

@larsenwork: this should fix many of the issues you encountered with your font (e.g. #3821), so I'd :heart: to have your feedback on this as well. You should be able to test drive it by:

* Cloning this repository;
* Checking out this branch;
* Following the [build instructions for your platform](https://github.com/atom/atom/tree/master/docs/build-instructions);
* Running the atom version you just built;

Thanks!

as-cii committed:
Fix a :bug: where some invalid lines were being measured

as-cii commented on src/text-editor-presenter.coffee:
```diff
@@ -595,33 +641,24 @@ class TextEditorPresenter
-      lastBufferRow = @model.bufferRowForScreenRow(rowBeforeStartRow)
-    else
-      lastBufferRow = null
+    @model.bufferRowForScreenRow(screenRow - 1) is bufferRow
```
@nathansobo: I feel like we should be able to optimize this. Instead of converting screen rows, do you think we could cache this information into `TokenizedLine` when we soft wrap a line in `DisplayBuffer`? This call is quite expensive in terms of performance, and it grows as the file gets longer.

Until we implement a faster `RowMap` I feel like we should be restrictive about calling those methods.

as-cii replied:
I went ahead and implemented this in 116f92d, as it seemed quite straightforward. Please, let me know if you think this could have any unintended side effects.

benogle commented:
> shouldn't notice any change from master, exception made for ligatures and kerning which should now work properly.

:boom:

as-cii committed:
Slightly more precise conversion to screen positions

alexandernst commented:
Extremely nice to see this happening. One little question: will ligatures work out-of-the-box or will there be some kind of option in the settings/change in the config file that should be enabled/made for ligatures to work?

as-cii committed:
:art: Move pixel-related code to LinesYardstick

* :fire: Remove useless 'scoped char width' methods

as-cii committed:
:racehorse: Mark soft-wrapped lines in TokenizedLine

izuzak commented:
I'm seeing this exception when testing this branch:

<img width="1280" alt="screen shot 2015-10-01 at 11 58 15" src="https://cloud.githubusercontent.com/assets/38924/10218060/6775f9a0-6834-11e5-80de-6f05be84410a.png">

Step to reproduce:

1. Copy this text

  ```
क ख ग घ ङ च छ ज झ ञ ट ठ ड ढ ण त थ द ध न प फ ब भ म य र ऱ ल ळ व श ष स ह क़ ख़ ग़ ज़ ड़ ढ़ फ़ य़ ॹ ॻ ॼ ॾ ॿ
```

2. Open a new tab in Atom
3. Paste the text a few times.
4. Use the mouse to position the cursor near the beginning of line 2
5 Press `delete` a few times so that line 2 merges with line 1
6. Select all text (`cmd-a`)
7. Press `delete`

GIF:

![exc2](https://cloud.githubusercontent.com/assets/38924/10218511/7f31d160-6837-11e5-8b44-1321f0ef7f45.gif)


Wasn't able to trigger it on `master`. Are you seeing the same, @as-cii?

izuzak commented:
@as-cii Also, do you know which issues in this repository this PR should fix (other than #3821), so that we can perhaps verify/test that this PR fixes those issues? :pray:

izuzak commented:
I was testing to see how #3821 worked now (Hasklig font + `text-rendering: optimizeLegibility;`), and noticed this behavior:

![exec3](https://cloud.githubusercontent.com/assets/38924/10218642/bde7a352-6838-11e5-89ef-184bcf2f3c65.gif)

When I try to go over the `===` character from right side to left side, the cursor stays in place for the first two key presses, and then goes over the character on the third key press (notice how the position of the cursor changes in the status bar). After that, I hop over to the right side again, press `delete` and the cursor jumps back in front of the character.

I'm not sure if this PR is supposed to address this or if this is even a problem, but wanted to point it out as something I noticed. :v:

izuzak commented:
Also notice this problem with the cursor vanishing:

![exec4](https://cloud.githubusercontent.com/assets/38924/10218751/7995ae6e-6839-11e5-8f98-715cdf45fd34.gif)

I type `======` then press `left`.

as-cii committed:
:bug: Exclude invalid tiles

as-cii committed:
:bug: Fix wrong pixel to screen position conversion

as-cii committed:
:green_heart:

as-cii commented:
First off, thanks @izuzak for the thorough feedback! :pray: 

So, most of my tests were performed using `Monoid`, which seemingly works as expected:

![monoid](https://cloud.githubusercontent.com/assets/482957/10219574/600909c6-6840-11e5-8b76-6d3e841991df.gif)

On the other hand, Hasklig behaves differently (i.e. exactly as @izuzak pointed out) also on Chrome Canary (e.g. tests performed on a `<textarea>`):

![hasklig](https://cloud.githubusercontent.com/assets/482957/10219676/3c5272be-6841-11e5-9b1c-510c2d562d6d.gif)

As you can see, the cursor cannot jump between physical characters. I still have to understand why we have this substantial difference between the two fonts (@larsenwork: maybe you can tell us more about this?), but I am not sure we can do much better than this unless we start to insert "breaks" in the HTML (which is a completely different beast, I think).

Nevertheless, and aside from the cursor disappearing (I'll fix this in a while), I still think this is a huge improvement over **master**, where fonts like Hasklig or Monoid or even subpixel font scaling messed up measurements on the whole line:

![master](https://cloud.githubusercontent.com/assets/482957/10219827/9d273240-6842-11e5-8bd5-561ba188b384.gif)

What do you think?

as-cii committed:
:bug: Fix an issue where yardstick cache wasn't being emptied

as-cii committed:
:racehorse: Avoid recomputing state for lines twice

izuzak commented:
> Nevertheless, and aside from the cursor disappearing (I'll fix this in a while), I still think this is a huge improvement over master, 

Definitely -- it is a huge improvement :+1: :purple_heart:. 

Let me know when you're done with those changes -- I'd be happy to test this PR one more time.

as-cii committed:
:white_check_mark: Thoroughly test LinesYardstick interactions

as-cii commented:
> Let me know when you're done with those changes -- I'd be happy to test this PR one more time.

The hidden cursor issue is a tricky one, as it's kinda difficult to reproduce it through a test case. I'll gather some more insights (also about the Hasklig font) tomorrow: in the meantime you can already try this out as every other issue should be fixed and your super helpful feedback would be truly appreciated, @izuzak. :sparkles:

izuzak commented:
> in the meantime you can already try this out as every other issue should be fixed and your super helpful feedback would be truly appreciated, @izuzak.

:cool: sounds good. Here's one I found on this branch (latest commit) but couldn't reproduce on master:

![lorem](https://cloud.githubusercontent.com/assets/38924/10232443/198316c4-6889-11e5-8b7c-2f8672c28510.gif)

Notice how when I trigger soft-wrap -- the width of the window isn't updated and I can still scroll.

izuzak commented:
Sorry, I should have mentioned this in the comment above: the soft wrap problem is present when I use the Hasklig font and `text-rendering: optimizeLegibility;`. It is _not_ present when I use the default font and remove `text-rendering: optimizeLegibility;` from my styles.less file.

as-cii committed:
:racehorse: Remove rows to measure in the next frame

as-cii committed:
:bug: Correctly measure RTL characters

This will fix also the "hidden cursor" issue we were experiencing.

/cc: @izuzak

as-cii committed:
Enable subpixel font scaling

For certain font sizes, enabling `textRendering: optimizeLegibility` caused a
bunch of measurement-related issues. You can reproduce it by setting the following in your stylesheet:

```
atom-text-editor {
  font-size: 14px;
  text-rendering: optimizeLegibility;
}
```

Although I wanted to defer subpixel font scaling to a later moment, it seems
like Chrome needs to have it enabled in order to properly support the
"legibility" path for text rendering. (I guess this is part of the reason why
the Chromium team enabled it by default at some point in the past.)

as-cii commented:
In the latest commit I have introduced subpixel font scaling: I wanted to defer this to a later moment, but this is mandatory if we want to fully support `optimizeLegibility`. Specs are :x: because `wrap-guide` and `autocomplete-plus` need to be upgraded in order to support subpixel precision (I have already opened two PRs to do that).

Also, please note that with this PR we will fix cursor position when there are RTL characters (i.e. arabic):

![arabic](https://cloud.githubusercontent.com/assets/482957/10252376/6035b898-6937-11e5-8a60-d976a85ea26b.gif)

This, of course, doesn't mean we provide a complete RTL editing experience, but I feel like this is a nice improvement towards it.

@izuzak: all the other issues (i.e. hidden cursor, weird soft-wrapping behavior) should have been fixed, so it would be awesome if you could give this another try! :pray:

izuzak commented:
> @izuzak: all the other issues (i.e. hidden cursor, weird soft-wrapping behavior) should have been fixed, so it would be awesome if you could give this another try!

:zap: @as-cii -- I couldn't find any other problems and confirmed those :point_up: were fixed.

nylki commented:
Has anybody tested if different font-sizes in one document behaves better with the new changes? Previously there where some wiered cursor positionings and line overlappings etc.

as-cii committed:
Merge branch 'master' into as-double-reflow-measurements

as-cii commented:
> I couldn't find any other problems and confirmed those  were fixed.

@izuzak: :green_heart: :heart: :purple_heart: :blue_heart:. If you happen to test this once more, it'd be great if you could confirm there are no performance regression and Atom is still fast. :zap: :racehorse: 

> Has anybody tested if different font-sizes in one document behaves better with the new changes? Previously there where some weired cursor positionings and line overlappings etc.

@nylki: as far as overlapping lines are concerned, this won't actually alter the existing behavior, but cursor positioning should improve with the changes in this PR. Maybe you could point out at some specific scenarios you had in mind so that we can test them out as well? (bonus points :sparkles: if you could give some feedback by building it locally :heart:)

Anyway, specs should now pass :green_heart: and I think we can :shipit: as soon as we have asserted there are no performance regression (I haven't noticed anything so far) and code-wise this looks good.

Thank you!

nathansobo commented:
Because lines are grouped into tiles that are absolutely positioned by multiplying the start row of the tile by the height of a "typical" line, we'll still have issues dealing with varying line heights, which will still cause troubles with changing line heights. It's something we'll need to address separately and it could be a while until we're ready, but definitely open to supporting it in the long run.

pradyunsg commented:
:heart:

as-cii committed:
:fire: Remove dead code

as-cii committed:
:racehorse: Avoid to measure invisible lines when dragging

as-cii committed:
Avoid invalidating state when clearing screen rows to measure

as-cii committed:
:art:

as-cii committed:
:racehorse: Faster line number calculation

as-cii committed:
Merge branch 'master' into as-double-reflow-measurements

wismill commented:
Sounds very good! I am testing it with the amazing font [PragmataPro](http://www.fsd.it/fonts/pragmatapro.htm).

I highly recommend following the [instruction from Laurian](https://github.com/atom/atom/issues/3821#issuecomment-110017100) that deactivate the ligature on the current line, in order to see exactly on which character the cursor is positioned while moving over a ligature ([some ligatures](http://www.fsd.it/fonts/pragmatapro/OpenType_features.png) involve 3 characters).

A package could be very useful to manage `text-rendering: optimizeLegibility;` and the option to deactivate it on the current line: simple options on/off.

as-cii committed:
Revert ":racehorse: Faster line number calculation"

This reverts commit e04aef0af3a707c49bf1c48ab01b78d4c79e446f.

as-cii committed:
Prepare DOMElementPool to account for text nodes

as-cii committed:
Recycle text nodes in line numbers

as-cii committed:
Recycle text nodes in lines

as-cii committed:
Faster DOM removal

as-cii committed:
Recycle highlights

as-cii committed:
Cache built text nodes

as-cii committed:
Use cached text nodes instead of NodeIterator

as-cii committed:
:racehorse:

as-cii committed:
:fire:

nathansobo commented:
Bravo @as-cii. An incredible amount of work went into this PR, much of it not even visible in the commit history. This is the first big step toward good support for internationalization!

benogle commented:
:boom:

maxbrunsfeld commented:
:fireworks: This is huge!

50Wliu commented:
Yesss! :100:

kevinsawicki commented:
So awesome :+1: :+1: :+1:

rajadain commented:
Ligature support! Incredible! I've been waiting for this for more than a year!

dikiaap commented:
:metal:
